lib: add aborted() utility function (#46494)
nodejs/node: doc/api/util.md

channel.port2.postMessage(signal, [signal]);
```

## `util.aborted(signal, resource)`

<!-- YAML
added: REPLACEME
-->

> Stability: 1 - Experimental

* `signal` {AbortSignal}
* `resource` {any} Any non-null entity
* Returns: {Promise}

Listens to abort event on the provided `signal` and returns a promise which is
resolved if an abort event is triggered. The function cleans up any
event listeners depending on when the provided `resource` is garbage collected.

```js
const { aborted } = require('util');

const ac = new AbortController();

> const dependent = {};

Reviewer: I don't like this example since you're supposed to pass a resource in - the issue is all our resources already take signals :D so maybe a mock call like `const dependent = obtainSomethingAbortable()` and calling `dependent.abort()` in the listener?

Author: Ah ok xD, updating

channel.port2.postMessage(signal, [signal]);
```

## `util.aborted(signal, resource)`

<!-- YAML
added: REPLACEME
-->

> Stability: 1 - Experimental

* `signal` {AbortSignal}

> * `resource` {any} Any non-null entity

Reviewer 2: This really isn't `any`, correct? It needs to be an object

Reviewer 2: Also, it should be stressed here that the reference to `resource` is held weakly.

Author: In the validator I have allowed objects, arrays, functions 
so make the time {Object|Array|Function}

Author: updated could you check

* Returns: {Promise}

Listens to abort event on the provided `signal` and returns a promise which is
resolved if an abort event is triggered. The function is dependent on the `resource`
entity and will be resolved only if the `resource` is not removed

> from memory, when `resource` goes out of memory the

Reviewer 2: Nit: weird line wrapping in here

Author: Ah sorry must configure my editor 😅

Author: updated

promise shall remain pending and any event listeners attached to the `signal`
will be removed.

> ```js

Reviewer 2: Nit: separate cjs/mjs examples

Author: updated

Reviewer 3: ``` suggestion
```cjs
```

channel.port2.postMessage(signal, [signal]);
```

## `util.aborted(signal, resource)`

<!-- YAML
added: REPLACEME
-->

> Stability: 1 - Experimental

* `signal` {AbortSignal}
* `resource` {Object} Any non-null entity, reference to which is held weakly.
* Returns: {Promise}

Listens to abort event on the provided `signal` and

> returns a promise that resolves when the `signal` is

Reviewer 3: ```suggestion
returns a promise that is fulfilled when the `signal` is
```

> aborted. The function is dependent on the `resource` and
> the returned promise shall remain pending if `resource`
> goes out of memory. The function shall automatically
> cleanup any event listeners it attaches to `signal`.

Reviewer 3: "the function is dependent" and "the function shall cleanup" is a bit unclear which function we're talking about. I think the fact that we're using a event listener is more an implementation detail than something we should document, and if anything the fact that we clean up goes without saying, so I would remove the sentence altogether.

```suggestion
aborted. If the passed `resource` is garbage collected,
the returned promise shall remain pending indefinitely.
```

Author: Ah yes went a little to verbose 😅

> Stability: 1 - Experimental

* `signal` {AbortSignal}
* `resource` {Object} Any non-null entity, reference to which is held weakly.
* Returns: {Promise}

Listens to abort event on the provided `signal` and
returns a promise that resolves when the `signal` is
aborted. The function is dependent on the `resource` and
the returned promise shall remain pending if `resource`
goes out of memory. The function shall automatically
cleanup any event listeners it attaches to `signal`.

```js

> const { aborted } = require('util');

Reviewer 3: ```suggestion
const { aborted } = require('node:util');
```
